src/Components/Endpoints/test/WebRootComponentParametersTest.cs

```diff
@@ -0,0 +1,114 @@
+
+        // Assert
+        Assert.False(result);
+    }
+
+    [Fact]
+    public void WebRootComponentParameters_DefinitelyEquals_ReturnsFalse_ForMismatchedParameterNames()
+    {
+        // Arrange
+        var parameters1 = CreateParameters(new() { ["First"] = 123 });
+        var parameters2 = CreateParameters(new() { ["Second"] = 123 });
+
+        // Act
+        var result = parameters1.DefinitelyEquals(parameters2);
+
+        // Assert
+        Assert.False(result);
+    }
+
+    [Fact]
+    public void WebRootComponentParameters_DefinitelyEquals_ReturnsFalse_ForMismatchedParameterValues()
+    {
+        // Arrange
+        var parameters1 = CreateParameters(new() { ["First"] = 123 });
+        var parameters2 = CreateParameters(new() { ["First"] = 456 });
+
+        // Act
+        var result = parameters1.DefinitelyEquals(parameters2);
+
+        // Assert
+        Assert.False(result);
+    }
+
+    [Fact]
+    public void WebRootComponentParameters_DefinitelyEquals_ReturnsFalse_ForMismatchedParameterTypes()
+    {
+        // Arrange
+        var parameters1 = CreateParameters(new() { ["First"] = 123 });
+        var parameters2 = CreateParameters(new() { ["First"] = 123L });
+
+        // Act
+        var result = parameters1.DefinitelyEquals(parameters2);
+
+        // Assert
+        Assert.False(result);
+    }
+
+    public static readonly object[][] DefinitelyEqualParameterValues =
+    [
+        [123],
+        ["abc"],
+        [new { First = 123, Second = "abc" }],
+    ];
+
+    [Theory]
+    [MemberData(nameof(DefinitelyEqualParameterValues))]
+    public void WebRootComponentParameters_DefinitelyEquals_ReturnsTrue_ForSameParameterValues(object value)
+    {
+        // Arrange
+        var parameters1 = CreateParameters(new() { ["First"] = value });
+        var parameters2 = CreateParameters(new() { ["First"] = value });
+
+        // Act
+        var result = parameters1.DefinitelyEquals(parameters2);
+
+        // Assert
+        Assert.True(result);
+    }
+
+    [Fact]
+    public void WebRootComponentParameters_DefinitelyEquals_ReturnsTrue_ForEmptySetOfParameters()
+    {
+        // Arrange
+        var parameters1 = CreateParameters(new());
+        var parameters2 = CreateParameters(new());
+
+        // Act
+        var result = parameters1.DefinitelyEquals(parameters2);
+
+        // Assert
+        Assert.True(result);
+    }
+
+    private static WebRootComponentParameters CreateParameters(Dictionary<string, object> parameters)
+    {
+        var parameterView = ParameterView.FromDictionary(parameters);
+        var (parameterDefinitions, parameterValues) = ComponentParameter.FromParameterView(parameterView);
+        for (var i = 0; i < parameterValues.Count; i++)
+        {
+            // WebRootComponentParameters expects parameter values to be JsonElements.
+            var jsonElement = JsonSerializer.SerializeToElement(parameterValues[i]);
+            parameterValues[i] = jsonElement;
```

Comment:
This is a bit sketchy as it's duplicating some implementation inside the unit test, meaning that the unit test is not really testing what its name seems to imply.

However looking at the implementation of `WebRootComponentParameters`, it looks like it will actually throw if you ever pass it any value that isn't a `JsonElement`. If that's the case, I think the test pattern here is fine, since there isn't any other possible scenario than `JsonElement`.

But also if that is true, could we have a unit test asserting that all the parameter values *must* be `JsonElement` and it will throw if not? Then we know how it behaves in all cases.